Add Stable Diffusion

fchollet opened the pull request:
This is a working Stable Diffusion text to image model. It reuses a bunch of code from Divam's original implementation. All top-level models are full rewrites as Functional models.

All in all it's about 600 LOC. We can probably go below we further refactoring. The UNet is too declarative right now, it could be refactored to be a lot more concise.

We need to:

- Refactor the UNet (`DiffusionModel`) to be as elegant as possible
- Figure out what to do with `constants.py`
- Upload the new weights files somewhere
- Further cleanup (remove the temporary test code) and beautify the code
- Add some basic unit testing

bhack commented:
/cc @divamgupta @innat

LukeWood commented:
So cool, can't wait to have SD ready and in the package...

bhack commented:
As this is the first materialized example where we are handling NLP+VISION I don't know if it is interesting to re-introduce some of our related topics like: 

Quoting @chenmoneygithub
https://github.com/keras-team/keras/issues/16181#issuecomment-1064492160
> And you are totally correct, we need to agree on a repo for shared components. We think it makes sense to put things into Keras main repo if both KerasCV and KerasNLP would depend on them.

Me:
https://github.com/keras-team/keras-cv/discussions/52#discussioncomment-3457106 https://github.com/keras-team/keras-cv/pull/30#issuecomment-1008685239

LukeWood commented on keras_cv/models/generative/stable_diffusion/decoder.py:
```diff
@@ -0,0 +1,97 @@
+import tensorflow as tf
+from tensorflow import keras
+
+from __internal__.layers.padded_conv2d import PaddedConv2D
```
from keras_cv.models.generative.stable_diffusion.__internal__.layers.padded_conv2d

LukeWood commented on keras_cv/models/generative/stable_diffusion/diffusion_model.py:
```diff
@@ -0,0 +1,283 @@
+import tensorflow as tf
+from tensorflow import keras
+
+from __internal__.layers.padded_conv2d import PaddedConv2D
```
same in this file

LukeWood commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -0,0 +1,177 @@
+from tensorflow import keras
+
+from decoder import Decoder
+from diffusion_model import DiffusionModel
```
from keras_cv.models.generative.stable_diffusion.decoder import Decoder
from keras_cv.models.generative.stable_diffusion.diffusion_model import DiffusionModel
from keras_cv.models.generative.stable_diffusion.text_encoder import TextEncoder
from keras_cv.models.generative.stable_diffusion.clip_tokenizer import SimpleTokenizer
from keras_cv.models.generative.stable_diffusion.constants import _UNCONDITIONAL_TOKENS, _ALPHAS_CUMPROD

bhack replied:
```python
from keras_cv.models.generative.stable_diffusion.text_encoder import TextEncoder
from keras_cv.models.generative.stable_diffusion.clip_tokenizer import SimpleTokenizer
```

Are this not NLP classes?

LukeWood reviewed (changes requested):
Minor changes

bhack commented on keras_cv/models/generative/stable_diffusion/__internal__/layers/group_normalization.py:
```diff
@@ -0,0 +1 @@
+from tensorflow_addons.layers import GroupNormalization
```
We need to remove this and duplicate/porting/refactor it in keras-cv or keras repo (https://github.com/keras-team/keras-cv/issues/74)

LukeWood replied:
@ianstenbit is working on this for now, it’ll live in core Keras - temporarily as a CV internal API

tanzhenyu commented on keras_cv/models/generative/stable_diffusion/text_encoder.py:
```diff
@@ -0,0 +1,125 @@
+            [input_tokens, token_positions]
+        )
+        for _ in range(12):
+            x = CLIPEncoderLayer()(x)
```
A general question -- how much of those text-related encoder do we want to declare as private API?

LukeWood replied:
In my opinion all text APIs should be __internal__ - CLIP should eventually be in KerasNLP so we should make it clear this won’t live here long term.

fchollet replied:
Yeah, and a key point here is that the Stable Diffusion *cannot be repurposed for anything else*. Our only goal is to match the numerics needed to load the open-source weights (the checkpoint cost ~800k to produce so we won't retrain). We won't use standard components or make components from the model reusable. Even the conv2d we use is non-standard (`PaddedConv2D`).

bhack replied:
Will this really be the general approach to "fundation models" in keras-cv?

It seems really an outlier for the library modularity manifesto.

We have started a thread at https://github.com/keras-team/keras-cv/issues/488#issuecomment-1154998812

Unfortunately, between internal team meetings and public discussions we never manage to stay on the same page. As community we always live with an information asymmetry / debt

tanzhenyu commented on keras_cv/models/generative/stable_diffusion/text_encoder.py:
```diff
@@ -0,0 +1,125 @@
+        self.v_proj = keras.layers.Dense(self.embed_dim)
+        self.out_proj = keras.layers.Dense(self.embed_dim)
+
+    def _shape(self, tensor, sequence_length, batch_size):
```
maybe `_compute_???_shape`?

tanzhenyu commented on keras_cv/models/generative/stable_diffusion/diffusion_model.py:
```diff
@@ -0,0 +1,279 @@
+
+
+def gelu(x):
+    tanh_res = keras.activations.tanh(x * 0.7978845608 * (1 + 0.044715 * (x**2)))
```
numbers seem pretty magical :-)

fchollet replied:
This is just a polynomial approximation of gelu. Generally speaking there are a number of constants harcoded in the code, because the code is non-configurable beyond the arguments of `StableDiffusion` and `text_to_image`. Its only function is to load the original weights and match the original numerics.

tanzhenyu reviewed (commented):
Excited to see this!
A general comment -- are we getting the weights from somewhere, or we will provide a training script later?

LukeWood commented:
> Excited to see this!
> A general comment -- are we getting the weights from somewhere, or we will provide a training script later?

As of now they’re ported.  If I can setup LAOIN-5B (hoping to in the next few months…) we can retrain; LAOIN-5B was the dataset this model was original trained on.

LukeWood commented:
Right now CLIP weight downloading is not lazy:

<img width="717" alt="Screen Shot 2022-09-23 at 11 37 41 PM" src="https://user-images.githubusercontent.com/12191303/192083762-75c6faad-d1e3-453c-b57a-eeb6c7039c40.png">

LukeWood reviewed (changes requested):
Don't forget to add an entry to keras_cv/models/__init__.py!

LukeWood commented:
<img width="760" alt="Screen Shot 2022-09-23 at 11 56 58 PM" src="https://user-images.githubusercontent.com/12191303/192084670-28b9064f-06f6-498d-a505-5d03ad023e1a.png">

Managed to get it working though!

bhack commented:
> > Excited to see this!
> > A general comment -- are we getting the weights from somewhere, or we will provide a training script later?
> 
> As of now they’re ported. If I can setup LAOIN-5B (hoping to in the next few months…) we can retrain; LAOIN-5B was the dataset this model was original trained on.

Are we going to target the recent OpenCLIP releases?

https://laion.ai/blog/large-openclip/

bhack commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -0,0 +1,138 @@
+        # Decoding stage
+        decoded = self.decoder.predict_on_batch(latent)
+        decoded = ((decoded + 1) / 2) * 255
+        return np.clip(decoded, 0, 255).astype("uint8")
```
` tfnp.clip` ?

fchollet replied:
It's already a numpy array (`predict_on_batch()` returns np arrays)

bhack commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -0,0 +1,138 @@
+
+        # Encode prompt tokens + positions into a "context" vector
+        pos_ids = np.array(list(range(MAX_PROMPT_LENGTH)))[None].astype("int32")
+        pos_ids = np.repeat(pos_ids, batch_size, axis=0)
```
`tfnp.repeat` ?

fchollet replied:
Same, this is numpy code already

bhack commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -0,0 +1,138 @@
+
+        # Encode unconditional tokens + positions as "unconditional context"
+        unconditional_tokens = np.array(_UNCONDITIONAL_TOKENS)[None].astype("int32")
+        unconditional_tokens = np.repeat(unconditional_tokens, batch_size, axis=0)
```
Ditto

bhack commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -0,0 +1,138 @@
+import numpy as np
```
Are we sure that we want to handle all these numpy objects and numpy TF experimental wrapped TF ops instead of handling directly Tensor and TF obs?

bhack replied:
I am asking this also for coherence:

```bash
grep -r "import numpy" * --exclude="*test.py" --exclude-dir="*examples*" --exclude-dir="*test*" --exclude-dir="*benchmarks*"
```
```bash
keras_cv/layers/preprocessing/random_rotation.py:import numpy as np
keras_cv/models/object_detection/__internal__.py:import numpy as np
keras_cv/models/object_detection/retina_net/retina_net.py:import numpy as np
```

Then also in these few cases it was used only for `np.log` and `np.pi`. 
(I don't know if there was a really rationale cause we could not use pi from python `math`/` tf.math.log`)

LukeWood commented on keras_cv/models/generative/stable_diffusion/clip_tokenizer.py:
```diff
@@ -0,0 +1,154 @@
+
+
+class SimpleTokenizer:
+    def __init__(self, bpe_path: str = default_bpe()):
```
here's the bug causing your lazy loading to not work:

```
bpe_path=None

bpe_path = bpe_path or default_bpe()
```

YaoYuanxin commented:
Hi, I have a fairly basic question. I installed the updated version of keras_cv, but it's showing "no attribute".
![image](https://user-images.githubusercontent.com/78332769/192121115-db008fc7-96d0-49eb-bc1d-18c846e6a1ca.png)

Is it because StableDiffusion hasn't been officially released for keras_cv yet?

LukeWood commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -126,13 +141,3 @@ def get_initial_parameters(self, timesteps, batch_size, seed=None):
             (batch_size, self.img_height // 8, self.img_width // 8, 4), seed=seed
         )
         return noise, alphas, alphas_prev
-
```
Feel free to add this to examples/models/stable_diffusion/

right now they’re basically glorified debugging demos; but we could render them someday down the line with proper polish

LukeWood commented:
> Hi, I have a fairly basic question. I installed the updated version of keras_cv, but it's showing "no attribute". ![image](https://user-images.githubusercontent.com/78332769/192121115-db008fc7-96d0-49eb-bc1d-18c846e6a1ca.png)
> 
> Is it because StableDiffusion hasn't been officially released for keras_cv yet?

Correct

LukeWood commented on keras_cv/models/generative/stable_diffusion/__internal__/layers/group_normalization.py:
```diff
@@ -0,0 +1 @@
+from tensorflow_addons.layers import GroupNormalization
```
lets just paste: 

```
# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Orginal implementation from keras_contrib/layer/normalization
# =============================================================================

import logging
import tensorflow as tf
from typeguard import typechecked

from tensorflow_addons.utils import types


@tf.keras.utils.register_keras_serializable(package="Addons")
class GroupNormalization(tf.keras.layers.Layer):
    """Group normalization layer.
    Source: "Group Normalization" (Yuxin Wu & Kaiming He, 2018)
    https://arxiv.org/abs/1803.08494
    Group Normalization divides the channels into groups and computes
    within each group the mean and variance for normalization.
    Empirically, its accuracy is more stable than batch norm in a wide
    range of small batch sizes, if learning rate is adjusted linearly
    with batch sizes.
    Relation to Layer Normalization:
    If the number of groups is set to 1, then this operation becomes identical
    to Layer Normalization.
    Relation to Instance Normalization:
    If the number of groups is set to the
    input dimension (number of groups is equal
    to number of channels), then this operation becomes
    identical to Instance Normalization.
    Args:
        groups: Integer, the number of groups for Group Normalization.
            Can be in the range [1, N] where N is the input dimension.
            The input dimension must be divisible by the number of groups.
            Defaults to 32.
        axis: Integer, the axis that should be normalized.
        epsilon: Small float added to variance to avoid dividing by zero.
        center: If True, add offset of `beta` to normalized tensor.
            If False, `beta` is ignored.
        scale: If True, multiply by `gamma`.
            If False, `gamma` is not used.
        beta_initializer: Initializer for the beta weight.
        gamma_initializer: Initializer for the gamma weight.
        beta_regularizer: Optional regularizer for the beta weight.
        gamma_regularizer: Optional regularizer for the gamma weight.
        beta_constraint: Optional constraint for the beta weight.
        gamma_constraint: Optional constraint for the gamma weight.
    Input shape:
        Arbitrary. Use the keyword argument `input_shape`
        (tuple of integers, does not include the samples axis)
        when using this layer as the first layer in a model.
    Output shape:
        Same shape as input.
    """

    @typechecked
    def __init__(
        self,
        groups: int = 32,
        axis: int = -1,
        epsilon: float = 1e-3,
        center: bool = True,
        scale: bool = True,
        beta_initializer: types.Initializer = "zeros",
        gamma_initializer: types.Initializer = "ones",
        beta_regularizer: types.Regularizer = None,
        gamma_regularizer: types.Regularizer = None,
        beta_constraint: types.Constraint = None,
        gamma_constraint: types.Constraint = None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.supports_masking = True
        self.groups = groups
        self.axis = axis
        self.epsilon = epsilon
        self.center = center
        self.scale = scale
        self.beta_initializer = tf.keras.initializers.get(beta_initializer)
        self.gamma_initializer = tf.keras.initializers.get(gamma_initializer)
        self.beta_regularizer = tf.keras.regularizers.get(beta_regularizer)
        self.gamma_regularizer = tf.keras.regularizers.get(gamma_regularizer)
        self.beta_constraint = tf.keras.constraints.get(beta_constraint)
        self.gamma_constraint = tf.keras.constraints.get(gamma_constraint)
        self._check_axis()

    def build(self, input_shape):

        self._check_if_input_shape_is_none(input_shape)
        self._set_number_of_groups_for_instance_norm(input_shape)
        self._check_size_of_dimensions(input_shape)
        self._create_input_spec(input_shape)

        self._add_gamma_weight(input_shape)
        self._add_beta_weight(input_shape)
        self.built = True
        super().build(input_shape)

    def call(self, inputs):

        input_shape = tf.keras.backend.int_shape(inputs)
        tensor_input_shape = tf.shape(inputs)

        reshaped_inputs, group_shape = self._reshape_into_groups(
            inputs, input_shape, tensor_input_shape
        )

        normalized_inputs = self._apply_normalization(reshaped_inputs, input_shape)

        is_instance_norm = (input_shape[self.axis] // self.groups) == 1
        if not is_instance_norm:
            outputs = tf.reshape(normalized_inputs, tensor_input_shape)
        else:
            outputs = normalized_inputs

        return outputs

    def get_config(self):
        config = {
            "groups": self.groups,
            "axis": self.axis,
            "epsilon": self.epsilon,
            "center": self.center,
            "scale": self.scale,
            "beta_initializer": tf.keras.initializers.serialize(self.beta_initializer),
            "gamma_initializer": tf.keras.initializers.serialize(
                self.gamma_initializer
            ),
            "beta_regularizer": tf.keras.regularizers.serialize(self.beta_regularizer),
            "gamma_regularizer": tf.keras.regularizers.serialize(
                self.gamma_regularizer
            ),
            "beta_constraint": tf.keras.constraints.serialize(self.beta_constraint),
            "gamma_constraint": tf.keras.constraints.serialize(self.gamma_constraint),
        }
        base_config = super().get_config()
        return {**base_config, **config}

    def _reshape_into_groups(self, inputs, input_shape, tensor_input_shape):

        group_shape = [tensor_input_shape[i] for i in range(len(input_shape))]
        is_instance_norm = (input_shape[self.axis] // self.groups) == 1
        if not is_instance_norm:
            group_shape[self.axis] = input_shape[self.axis] // self.groups
            group_shape.insert(self.axis, self.groups)
            group_shape = tf.stack(group_shape)
            reshaped_inputs = tf.reshape(inputs, group_shape)
            return reshaped_inputs, group_shape
        else:
            return inputs, group_shape

    def _apply_normalization(self, reshaped_inputs, input_shape):

        group_shape = tf.keras.backend.int_shape(reshaped_inputs)
        group_reduction_axes = list(range(1, len(group_shape)))
        is_instance_norm = (input_shape[self.axis] // self.groups) == 1
        if not is_instance_norm:
            axis = -2 if self.axis == -1 else self.axis - 1
        else:
            axis = -1 if self.axis == -1 else self.axis - 1
        group_reduction_axes.pop(axis)

        mean, variance = tf.nn.moments(
            reshaped_inputs, group_reduction_axes, keepdims=True
        )

        gamma, beta = self._get_reshaped_weights(input_shape)
        normalized_inputs = tf.nn.batch_normalization(
            reshaped_inputs,
            mean=mean,
            variance=variance,
            scale=gamma,
            offset=beta,
            variance_epsilon=self.epsilon,
        )
        return normalized_inputs

    def _get_reshaped_weights(self, input_shape):
        broadcast_shape = self._create_broadcast_shape(input_shape)
        gamma = None
        beta = None
        if self.scale:
            gamma = tf.reshape(self.gamma, broadcast_shape)

        if self.center:
            beta = tf.reshape(self.beta, broadcast_shape)
        return gamma, beta

    def _check_if_input_shape_is_none(self, input_shape):
        dim = input_shape[self.axis]
        if dim is None:
            raise ValueError(
                "Axis " + str(self.axis) + " of "
                "input tensor should have a defined dimension "
                "but the layer received an input with shape " + str(input_shape) + "."
            )

    def _set_number_of_groups_for_instance_norm(self, input_shape):
        dim = input_shape[self.axis]

        if self.groups == -1:
            self.groups = dim

    def _check_size_of_dimensions(self, input_shape):

        dim = input_shape[self.axis]
        if dim < self.groups:
            raise ValueError(
                "Number of groups (" + str(self.groups) + ") cannot be "
                "more than the number of channels (" + str(dim) + ")."
            )

        if dim % self.groups != 0:
            raise ValueError(
                "Number of groups (" + str(self.groups) + ") must be a "
                "multiple of the number of channels (" + str(dim) + ")."
            )

    def _check_axis(self):

        if self.axis == 0:
            raise ValueError(
                "You are trying to normalize your batch axis. Do you want to "
                "use tf.layer.batch_normalization instead"
            )

    def _create_input_spec(self, input_shape):

        dim = input_shape[self.axis]
        self.input_spec = tf.keras.layers.InputSpec(
            ndim=len(input_shape), axes={self.axis: dim}
        )

    def _add_gamma_weight(self, input_shape):

        dim = input_shape[self.axis]
        shape = (dim,)

        if self.scale:
            self.gamma = self.add_weight(
                shape=shape,
                name="gamma",
                initializer=self.gamma_initializer,
                regularizer=self.gamma_regularizer,
                constraint=self.gamma_constraint,
            )
        else:
            self.gamma = None

    def _add_beta_weight(self, input_shape):

        dim = input_shape[self.axis]
        shape = (dim,)

        if self.center:
            self.beta = self.add_weight(
                shape=shape,
                name="beta",
                initializer=self.beta_initializer,
                regularizer=self.beta_regularizer,
                constraint=self.beta_constraint,
            )
        else:
            self.beta = None

    def _create_broadcast_shape(self, input_shape):
        broadcast_shape = [1] * len(input_shape)
        is_instance_norm = (input_shape[self.axis] // self.groups) == 1
        if not is_instance_norm:
            broadcast_shape[self.axis] = input_shape[self.axis] // self.groups
            broadcast_shape.insert(self.axis, self.groups)
        else:
            broadcast_shape[self.axis] = self.groups
        return broadcast_shape
```

in here - I think we can skip testing here and visually test.

ianstenbit commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -0,0 +1,143 @@
+MAX_PROMPT_LENGTH = 77
+
+
+class StableDiffusion:
```
Since we don't have a tutorial yet, maybe a code example in a docstring here would be nice?

ianstenbit replied:
(especially since this code automagically loads weights, which might be a bit confusing to users)

fchollet replied:
Good call -- I added a full docstring (and an example in `examples/...`)

ianstenbit commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -0,0 +1,143 @@
+        text_encoder_weights_fpath = keras.utils.get_file(
+            origin="https://huggingface.co/fchollet/stable-diffusion/resolve/main/kcv_encoder.h5",
+            file_hash="4789e63e07c0e54d6a34a29b45ce81ece27060c499a709d556c7755b42bb0dc4",
+        )
```
Should we let users load these weights from a custom directory so that they can have them stored locally if they'd like (instead of making a big network call every time a model is instantiated)

fchollet replied:
The weights are cached automatically on local disk. We could add an option to specify the cache directory though! I'll leave that for a future update. Right now the cache is the default, at `~/.keras/datasets`.

fchollet commented:
I added a docstring, removed the TFA dependency, added a code example, and disabled the golden value test so that it doesn't run on CI (which would be overly expensive -- we don't want to download the weights on CI).

I believe we're ready to go, or pretty close.

LukeWood commented:
/gcbrun

divamgupta commented:
There will be a lot of transformer based models released in the future. Should there by separate modules for transformers?

LukeWood reviewed (approved):
I have reviewed the full PR, and this LGTM.  I have also pulled locally and played with the implementation; great job on the readability and organization.

LukeWood commented on keras_cv/models/__init__.py:
```diff
@@ -38,6 +38,7 @@
 from keras_cv.models.efficientnet_v2 import EfficientNetV2L
```
copyright in this one

LukeWood commented on keras_cv/models/generative/stable_diffusion/constants.py:
```diff
@@ -0,0 +1,1081 @@
+_UNCONDITIONAL_TOKENS = [
```
copyright here

LukeWood commented:
@fchollet you also need to add `"regex"` to `install_requires` in setup.py

fchollet commented:
Somehow I'm able to run `format.sh` and `lint.sh` locally just fine, yet it fails on CI.

LukeWood commented:
/gcbrun

LukeWood commented:
Adding a dependency to our GCB cluster; one monent...

LukeWood commented:
/gcbrun

LukeWood commented:
All checks pass @fchollet !  Thank you for working so swiftly to prepare this port!  Merging and rebasing https://github.com/keras-team/keras-cv/pull/831 on top.

Excited to have such a powerful offering in our package; and can't wait to show off the performance in the keras.io guide.
